images/controller/internal/controllers/rvr_diskful_count/reconciler.go

```diff
@@ -18,39 +18,321 @@ package rvrdiskfulcount
+			log.Info("ReplicatedVolume not found, ignoring reconcile request")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("getting ReplicatedVolume: %w", err)
+	}
+
+	if rv.DeletionTimestamp != nil {
+		log.Info("ReplicatedVolume is being deleted, ignoring reconcile request")
+		return reconcile.Result{}, nil
+	}
+
+	// Get diskful replica count
+	neededNumberOfReplicas, err := getDiskfulReplicaCount(ctx, r.cl, rv)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("getting diskful replica count: %w", err)
+	}
+	log.V(4).Info("Calculated diskful replica count", "count", neededNumberOfReplicas)
+
+	// Get all RVRs for this RV
+	totalRvrMap, err := getDiskfulReplicatedVolumeReplicas(ctx, r.cl, rv)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("getting ReplicatedVolumeReplicas: %w", err)
+	}
+	totalNumberOfReplicas := len(totalRvrMap)
+	log.V(4).Info("Found ReplicatedVolumeReplicas", "count", totalNumberOfReplicas)
+
+	// If no RVRs found, create one
+	if totalNumberOfReplicas == 0 {
+		log.Info("No ReplicatedVolumeReplicas found for ReplicatedVolume, creating one")
+		err = createReplicatedVolumeReplica(ctx, r.cl, r.scheme, rv, log)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("creating ReplicatedVolumeReplica: %w", err)
+		}
+
+		err = setDiskfulReplicaCountReachedCondition(
+			ctx, r.cl, log, rv,
+			metav1.ConditionFalse,
+			"FirstReplicaIsBeingCreated",
+			fmt.Sprintf("Created first replica, need %d diskful replicas", neededNumberOfReplicas),
+		)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("setting DiskfulReplicaCountReached condition: %w", err)
+		}
+
+		return reconcile.Result{}, nil
+	}
+
+	deletedRvrMap := getDeletedReplicas(totalRvrMap)
+	deletedNumberOfReplicas := len(deletedRvrMap)
+	log.V(4).Info("Counted deleting ReplicatedVolumeReplicas", "count", deletedNumberOfReplicas)
+
+	// Count non-deleted ReplicatedVolumeReplicas
+	nonDeletedNumberOfReplicas := totalNumberOfReplicas - deletedNumberOfReplicas
+	log.V(4).Info("Counted non-deleted ReplicatedVolumeReplicas", "count", nonDeletedNumberOfReplicas)
+
+	if nonDeletedNumberOfReplicas == 0 {
+		log.Info("No non-deleted ReplicatedVolumeReplicas found for ReplicatedVolume, creating one")
+		err = createReplicatedVolumeReplica(ctx, r.cl, r.scheme, rv, log)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("creating ReplicatedVolumeReplica: %w", err)
+		}
+
+		err = setDiskfulReplicaCountReachedCondition(
+			ctx, r.cl, log, rv,
+			metav1.ConditionFalse,
+			"FirstReplicaIsBeingCreated",
+			fmt.Sprintf("Created non-deleted replica, need %d diskful replicas", neededNumberOfReplicas),
+		)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("setting DiskfulReplicaCountReached condition: %w", err)
+		}
+
+		return reconcile.Result{}, nil
+	}
+
+	// Need to wait until RVR becomes Ready.
+	if nonDeletedNumberOfReplicas == 1 {
+		for _, rvr := range deletedRvrMap {
```

Comment:
Here we check nonDeletedNumberOfReplicas == 1, but then iterate over deletedRvrMap.
In the normal case (one non-deleted replica, no deleted replicas), deletedRvrMap is empty, the loop never runs, and the reconciler continues as if the first replica is already ready. As a result, additional replicas can be created even when the first replica is not Ready, which violates the requirement that the first diskful replica must reach a fully operational state before creating the second one.

We should instead check the non-deleted RVR(s), e.g. iterate over all replicas and skip those with DeletionTimestamp != nil, or introduce a separate getNonDeletedReplicas helper and use it both for counting and readiness checks.